### PR TITLE
Acquire token for connectors and server

### PR DESCRIPTION
**Pull request by teresaqhoang**
### Motivation and Context

This PR adds support to acquire an access token using MSAL. It passes this under a new header when making SK server requests, and the user ID token is also added under the 'Authorization' header.

### Description

This implementation follows the 'incremental consent' concept, where the User will only be prompted for 'openid', 'offline_access', 'profile', and 'User.Read' consent on login. Further permissions will require additional consent, in which the user will be prompted when a request for the token is made. 

If consent is granted and user is authorized, the access token will be passed to the server under this header: `sk-copilot-connector-access-token`. The `Authorization` header is reserved for the SK Access token.

To invoke skills with tokens, do the following (see comments in useChat for example):

1. `import { useConnectors } from '{path}/connectors/useConnectors';`
2. In component: `const connectors = useConnectors();`
3. Invoke Skill with scopes required for connector as an array, i.e. `const scopes = ['User.Read']`
To use Graph token specifically, comment out the scopes you need under `msGraphScopes` in `Constants.ts`, then call the i.e., `invokeSkillWithGraphToken` function. Be default, the ones already uncommented map to Graph APIs used in existing connectors.
`var result = await connectors.invokeSkillWithGraphToken(ask, 'ChatSkill', 'Chat');`
To use ADO token specifically, comment out the scopes you need under `adoScopes` in `Constants.ts`, then call the `invokeSkillWithAdoToken` function.
i.e., `var result = await connectors.invokeSkillWithAdoToken(ask, 'ChatSkill', 'Chat');`

+ Error Handling

### Contribution Checklist

- [X] The code builds clean without any errors or warnings
- [X] The PR follows SK Contribution Guidelines (https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md)
- [X] The code follows the .NET coding conventions (https://learn.microsoft.com/dotnet/csharp/fundamentals/coding-style/coding-conventions) verified with `dotnet format`
- [X] All unit tests pass, and I have added new tests where possible
- [X ] I didn't break anyone :smile:

### CODE REVIEW

File: samples/apps/copilot-chat-app/WebApp/src/Constants.ts [hunk 1]
```diff
@@ -66,4 +33,43 @@ export const Constants = {
             defaultDefinition: 'int',
         },
     },
+    // NOT a comprehensize list.
```
Comment (glahaye) [inline samples/apps/copilot-chat-app/WebApp/src/Constants.ts:36]: comprehensize -> comprehensive

File: samples/apps/copilot-chat-app/WebApp/src/libs/connectors/useConnectors.ts [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+                const request = new URL('/v1.0' + api, 'https://graph.microsoft.com');
+                return fetch(request, {
+                    method: method,
+                    headers: {
+                        ...apiHeaders,
+                        Authorization: `Bearer ${token}`,
+                    },
+                }).then(async (response) => {
+                    if (!response || !response.ok) {
+                        throw new Error(`Recieved error while request ${api}: ${response}`);
```
Comment (glahaye) [inline samples/apps/copilot-chat-app/WebApp/src/libs/connectors/useConnectors.ts:27]: Recieved -> Received

File: samples/apps/copilot-chat-app/WebApp/src/libs/semantic-kernel/SemanticKernel.ts [hunk 3]
```diff
@@ -13,39 +15,61 @@ export class SemanticKernel {
...
+                method: 'POST',
+                body: ask,
+            },
+            connectorAccessToken,
+        );
         return result;
     };
 
-    private readonly getResponseAsync = async <T>(request: ServiceRequest): Promise<T> => {
+    private readonly getResponseAsync = async <T>(
```
Comment (glahaye) [inline samples/apps/copilot-chat-app/WebApp/src/libs/semantic-kernel/SemanticKernel.ts:35]: So, SK as a service is, well, a service.  As such, it requires an access token to access 

Using an ID token to authenticate is not the best in terms of security: https://auth0.com/blog/amp/id-token-access-token-what-is-the-difference/

Also, the middleware in ASP.Net will not authorize endpoints using ID tokens. I'd have to write my own authorizer to make that happen 

To authorize properly, I'll just need a basic access token (probably one with only the "access_as_user" scope)

**Review by glahaye — CHANGES_REQUESTED** [review]
Please send an access token to SKaaS

File: samples/apps/copilot-chat-app/WebApp/src/libs/useChat.ts [hunk 4]
```diff
@@ -59,14 +63,16 @@ export const useChat = () => {
         const ask = { input: value, variables: [{ key: 'audience', value: account?.name ?? 'Unknown User' }] };
         try {
             var result = await sk.invokeAsync(ask, 'ChatSkill', 'Chat');
+            // var result = await connectors.invokeSkillWithGraphToken(ask, 'ChatSkill', 'Chat');
```
Comment (TaoChenOSU) [inline samples/apps/copilot-chat-app/WebApp/src/libs/useChat.ts:66]: Why do we need the graph token for Chat? Is this just an example statement?

Comment (teresaqhoang) [reply samples/apps/copilot-chat-app/WebApp/src/libs/useChat.ts:66]: Yeah this is just an example, I can change the skill to OpenAI Skill to make it more clear

File: samples/apps/copilot-chat-app/WebApp/src/libs/semantic-kernel/SemanticKernel.ts (hunk 3 above)
Comment (teresaqhoang) [reply samples/apps/copilot-chat-app/WebApp/src/libs/semantic-kernel/SemanticKernel.ts:35]: Changed to pass Access Token with initial scopes as defined by skScopes in Constants.ts. Will continue discussion offline about Server Auth Story

File: samples/apps/copilot-chat-app/WebApp/src/libs/useChat.ts (hunk 4 above)
Comment (teresaqhoang) [reply samples/apps/copilot-chat-app/WebApp/src/libs/useChat.ts]: Changed to be more generic to ConnectorSkill. Added some helpful sections in ReadMe as well

File: samples/apps/copilot-chat-app/WebApp/README.md [hunk 5]
```diff
@@ -41,6 +41,35 @@ The Copilot Chat sameple showcases how to build an enriched intelligent app, wit
...
+
+To invoke skills with the required tokens, do the following (see comments tagged `ConnectorTokenExample` in [useChat](./src/libs/useChat.ts) for example):
+
+1. Import the Connectors hook: `import { useConnectors } from '{path}/connectors/useConnectors';`
+2. In component: `const connectors = useConnectors();`
+3. Invoke Skill with scopes required for the downstream Connector as an array, i.e. `const scopes = ['User.Read']`
+
+   `var result = await connectors.invokeSkillWithConnectorToken(ask, {ConnectorSkill}, {ConnectorFunction}, scopes);`
+
+      - To use Graph token specifically, comment out the scopes you need under `msGraphScopes` in [Constants.ts file](./src/Constants.ts), then call the `invokeSkillWithGraphToken` function. Be default, the ones already uncommented map to Graph APIs used in existing connectors.
```
Comment (glahaye) [inline samples/apps/copilot-chat-app/WebApp/README.md:68]: comment out?  You mean uncomment??

File: samples/apps/copilot-chat-app/WebApp/README.md [hunk 6]
```diff
@@ -41,6 +41,35 @@ The Copilot Chat sameple showcases how to build an enriched intelligent app, wit
 
 We need keys to work with various aspects of the project including accessing openAI models. This opens up the possibility of exposing keys in commits. There are a [couple of options](https://learn.microsoft.com/en-us/aspnet/core/security/app-secrets?view=aspnetcore-7.0&tabs=windows) to safeguard developers from exposing keys. Outside of using the dotnet's users-secrets and environment variables, we've also added *.development.json and *.development.config to the .gitignore if developers want to use appsettings.development.json files or other development.config files for secret storage.
```
Comment (glahaye) [inline samples/apps/copilot-chat-app/WebApp/README.md:42]: Nit: openAI -> OpenAI

File: samples/apps/copilot-chat-app/WebApp/README.md [hunk 7]
```diff
@@ -41,6 +41,35 @@ The Copilot Chat sameple showcases how to build an enriched intelligent app, wit
...
+2. In component: `const connectors = useConnectors();`
+3. Invoke Skill with scopes required for the downstream Connector as an array, i.e. `const scopes = ['User.Read']`
+
+   `var result = await connectors.invokeSkillWithConnectorToken(ask, {ConnectorSkill}, {ConnectorFunction}, scopes);`
+
+      - To use Graph token specifically, comment out the scopes you need under `msGraphScopes` in [Constants.ts file](./src/Constants.ts), then call the `invokeSkillWithGraphToken` function. Be default, the ones already uncommented map to Graph APIs used in existing connectors.
+      
+         i.e., `var result = await connectors.invokeSkillWithGraphToken(ask, {ConnectorSkill}, {ConnectorFunction});`
+
+      - To use ADO token specifically, comment out the scopes you need under `adoScopes` in [Constants.ts file](./src/Constants.ts), then call the `invokeSkillWithAdoToken` function.
```
Comment (glahaye) [inline samples/apps/copilot-chat-app/WebApp/README.md:72]: Again, comment out -> uncomment?

**Review by glahaye — DISMISSED** [review]

**Review by TaoChenOSU — APPROVED** [review]

**Review by glahaye — APPROVED** [review]

**Pull request merged**